### PR TITLE
(HSD 15011258925) Fix Kernel modules missing after AOTA update HDDL d…

### PR DESCRIPTION
**Pull request by kathashok**
The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
_Fix Kernel modules missing after AOTA update HDDL driver_

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): https://hsdes.intel.com/appstore/article/#/15011258925
- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [x] Changelogs are updated (or N/A if not customer visible)

### CODE MAINTAINABILITY
- [x] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [x] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [x] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):
- [x] Run 'go fmt' or autopep8 as applicable.
- [ ] New/modified methods and functions should have type annotations on signatures as applicable
- [ ] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [x] **_Quality of code (At least one should be checked as applicable)\*_**
	- [x] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [x] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [x] **_Static code scan report shows zero critical issues\*_**
- [x] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

### CODE REVIEW

File: inbm/Changelog.md [hunk 1]
```diff
@@ -7,7 +7,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
  - HSD 15011298374 - Missing AppArmor profile entries under /sys for diagnostic agent
-
+ - HSD 15011258925 - Kernel modules missing after AOTA update HDDL driver
+ - RTC 495463 - Existing driver get uninstalled if AOTA update failed using .deb file package
```
Comment (pillaiabhishek) [inline inbm/Changelog.md:11]: Remove the RTC reference.

File: inbm/dispatcher-agent/dispatcher/aota/application_command.py [hunk 2]
```diff
@@ -135,12 +151,23 @@ def update(self) -> None:
...
+            uninstall_driver_cmd = CHROOT_PREFIX + \
+                f'/usr/bin/rpm -e --nodeps {old_driver_name}'
+            out, err, code = PseudoShellRunner().run(uninstall_driver_cmd)
+
+            # If old packages wasn't install on system, it will return error too.
+            if code != 0 and "is not installed" not in str(err):
+                raise AotaError(err)
+
             chroot_driver_path = driver_centos_path.replace("/host", "")
             install_driver_cmd = CHROOT_PREFIX + \
```
Comment (pillaiabhishek) [inline inbm/dispatcher-agent/dispatcher/aota/application_command.py:167]: in the older RTC the install cmd that got replaced was "install_driver_cmd = DOCKER_CHROOT_PREFIX + f'**/usr/bin/rpm -ivh** {chroot_driver_path}'"

Are we not reverting back to this? I still see the -Uvh option being used.
Its probably not an issue as we are uninstalling the older version first and then calling the upgrade (-Uvh) cmd which may equate to an install (-ivh) cmd.

File: inbm/dispatcher-agent/dispatcher/aota/application_command.py [hunk 3]
```diff
@@ -135,12 +151,23 @@ def update(self) -> None:
...
                 f'/usr/bin/rpm -Uvh --oldpackage {chroot_driver_path}'
             logger.debug(f" Updating Driver {driver_path.split('/')[-1]} ...")
             out, err, code = PseudoShellRunner().run(install_driver_cmd)
-            logger.debug(out)
```
[the remark this answers is not in the document]
Comment (kathashok) [reply inbm/dispatcher-agent/dispatcher/aota/application_command.py:143]: After adding new changes in try block, we are getting build error like below.
**WPS229 Found too long ``try`` body length: 18 > 17 [wemake-python-styleguide]**

In ./inbm/common-python-config/pyproject.toml file max_try_body_length is defined as below.
max_try_body_length = 17 # very high

So either we need to update this value (from 17 to 18 or more) or we need to remove some lines in try block like log messages which are not important.

File: inbm/Changelog.md (hunk 1 above)
Comment (kathashok) [reply inbm/Changelog.md]: Removed in latest commit.

File: inbm/dispatcher-agent/dispatcher/aota/application_command.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nmgaston) [reply inbm/dispatcher-agent/dispatcher/aota/application_command.py:143]: I would put some or all of the code that you added into a separate method and call it.  Then it won't increase that count by much and you can leave the debug statement in.

File: inbm/dispatcher-agent/dispatcher/aota/application_command.py [hunk 4]
```diff
@@ -116,6 +116,16 @@ def cleanup(self) -> None:
         for file in os.listdir(CENTOS_DRIVER_PATH):
             remove_file(os.path.join(CENTOS_DRIVER_PATH, file))
 
+    def check_file_type(self, file_path: str) -> bool:
+        """Check the driver file is rpm type or not
+
+        @return: return False if file is not rpm type
+        """
+        if file_path.endswith('.rpm'):
```
[the remark this answers is not in the document]
Comment (kathashok) [reply inbm/dispatcher-agent/dispatcher/aota/application_command.py]: Updated as per comment. But getting build error (unit test case is failing) if we make new method as private.

[the remark this answers is not in the document]
Comment (kathashok) [reply inbm/dispatcher-agent/dispatcher/aota/application_command.py]: Updated is_rpm_file_type() method as private method in latest commit. Please check.

**Review by nmgaston — APPROVED** [review]

**Review by pillaiabhishek — APPROVED** [review]

**Review by gblewis1 — APPROVED** [review]

**Pull request merged**